src/driveshaft-config.cpp

```diff
@@ -0,0 +1,269 @@
+#include <fstream>
+#include <boost/range/algorithm/copy.hpp>
+#include <boost/range/adaptors.hpp>
+#include <boost/filesystem.hpp>
+#include "driveshaft-config.h"
+
+namespace Driveshaft {
+
+namespace cfgkeys {
+static std::string GEARMAN_SERVERS_LIST = "gearman_servers_list";
+static std::string POOLS_LIST = "pools_list";
+static std::string POOL_WORKER_COUNT = "worker_count";
+static std::string POOL_JOB_LIST = "jobs_list";
+static std::string POOL_JOB_PROCESSING_URI = "job_processing_uri";
+}
+
+DriveshaftConfig::DriveshaftConfig() noexcept :
+    m_config_filename(),
+    m_server_list(),
+    m_pool_map(),
+    m_load_time(0) {
+}
+
+bool DriveshaftConfig::load(const std::string &config_filename, Json::CharReader::Factory &parser_factory) {
+    if (!this->needsConfigUpdate(config_filename)) {
+        return false;
+    }
+
+    LOG4CXX_INFO(MainLogger, "Reloading config");
+    this->m_config_filename = config_filename;
+
+    std::string config_file_contents(this->fetchFileContents(config_filename));
+    return this->parseConfig(config_file_contents, parser_factory);
+}
+
+bool DriveshaftConfig::parseConfig(const std::string &config_data, Json::CharReader::Factory &parser_factory) {
+    Json::Value tree;
+    std::string parse_errors;
+    auto json_parser = parser_factory.newCharReader();
+    if (!json_parser->parse(config_data.data(),
+                            config_data.data() + config_data.length(),
+                            &tree, &parse_errors)) {
+        LOG4CXX_ERROR(MainLogger, "Failed to parse " << this->m_config_filename <<
+                                  ". Errors: " << parse_errors);
+        throw std::runtime_error("config parse failure");
+    }
+
+    if (!this->validateConfigNode(tree)) {
+        throw std::runtime_error("config parse failure");
+    }
+
+    this->m_load_time = std::time(nullptr);
```

Comment:
Here's one divergence - the prior code set m_load_time from the stat time on the config file. Due to the restructuring, that context is no longer available here. Semantically, using the current time may make more sense anyway, at least according to the variable's name. The only risk I associated with this change that I could identify is one where we could fail to identify two config reloads in very rapid (sub-second) succession. I would be very hard to convince that such a situation is worthy of real concern.